share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+
+## `BigArgumentsUsePointersToCopy` (`bool`)
+
+This allows a non-position-based ABI to take advantage of one of the main traits of position-based ABI: ability to use pointers to copy to pass big aggregates.
+
+This is one of the defining features of arm64 ABI, which allows it to get the best of both approaches - an ability to pass aggregates by copy in registers from the non-position-based approach while also being able to sometimes avoid a copy of a bigger aggregate from the position-based approach.
+
+## `NoRegisterArgumentsCanComeAfterStackOnes` (`bool`)
+
+States whether ABI allows a stack argument (mainly if it's too big to be placed in the registers) to precede other register arguments.
+
+For example, if there is a `Big` struct that has to use the stack, and
+a function like
+```cpp
+void function(Big, uint32_t);
+```
+if this value is set to true, both argument will be passed on stack, otherwise, only the struct will.
+
+## `AllowUnnaturallyAlignedTypesInRegisters` (`bool`)
+
+States whether ABI allows unnaturally aligned types[^1] to be passed in registers. When set to false, all the unnatural types will only be put on stack.
+
+## `CalleeIsResponsibleForStackCleanup` (`bool`)
+
+Specifies who is responsible for cleaning the stack after the function call. If equal to `true`, it's the callee, otherwise it the caller.
+
+## `FloatsUseGPRs` (`bool`)
+
+Setting this to true disables usage of vector registers entirely.
+
+## `StackArgumentsUseRegularStructAlignmentRules` (`bool`)
+
+Setting this to true allows stack arguments to use slots smaller than a word (makes them behave like a single struct instead of a bunch of arguments): so they don't get extended unnecessarily. A notable ABI using this is apple flavor of the aarch64.
+
+For example
+```cpp
+void function(/* enough arguments of all the registers */, uint8_t, uint8_t);
+```
+When this is set to true, both `uint8_t` will get placed as if they were a struct, that is without any padding in-between. When this is false, both of them will get extended to register size before getting placed, leading to padding.
+
+## `UseStrictAggregateAlignmentRules` (`bool`)
+
+Treat all the structs passed using the stack as if they were aligned unnaturally[^1].
+
+This means that they are going to handled with extreme care (which usually just means extra padding).
+
+## `StackAlignment` (`unsigned`)
+
+States the required alignment of the stack at the point of a function call in bytes.
+
+> **_Note:_** states minimum value for ABIs supporting multiple different stack alignment values, for example, if the ABI requires the stack to be aligned on 4 bytes for internal calls but on 8 bytes for interfaces (like 32-bit ARM ABI), the value of `StackAlignment` should be equal to 4.
+
+
+## `MinimumStackArgumentSize` (`unsigned`)
+
+States the minimum possible stack argument size in bytes. When the actual type of an argument is below this value, some padding with be inserted.
```

Comment:
I wonder how this plays with `StackArgumentsUseRegularStructAlignmentRules`.

Reply:
`StackArgumentsUseRegularStructAlignmentRules` is specifically added to ignore this size.

Reply:
Can't we merge the two options somehow? Maybe with a special value for `MinimumStackArgumentSize`?

> StackArgumentsUseRegularStructAlignmentRules is specifically added to ignore this size.

Since that's the case, state it in the docs.

I'm starting to thing that some grouping of these options (in the docs) would be helpful. Maybe we can have a section about stack arguments, a section about return values and things like this.

Reply:
> Can't we merge the two options somehow?

Would be pretty awkward, I think. Since it's not like `StackArgumentsUseRegularStructAlignmentRules` overrides `MinimumStackArgumentSize` per se - it's still there, kind of. It just allows multiple small arguments to get implicitly "struct"-ified.

> StackArgumentsUseRegularStructAlignmentRules is specifically added to ignore this size.

`*` in certain cases

(my bad on wording)

> grouping

Yeah, that was the original intention, but we only had a handful after the first iteration, so it got postponed. Eventually we'll have to get to it :D

As of now they are de-facto grouped like:
1. Name.
2. All the weird bools.
3. Register stuff.
4. Scalar type stuff.

Reply:
Can you reorganize the document in sections?

Reply:
Please take a look (it's in `squash 2`)

Reply:
Works, but drop the index, it will go out of sync in no time.

Reply:
It's generated by a vscode extension, so it's not going to get out of sync if I'm the one changing the options.

The question is how often do you expect other people adding removing options from here? Because if not very, I'd rather we keep the index.

Reply:
Frankly I see no reason why this shouldn't end up in docs.rev.ng. I'd add a "Developer's manual" section to `mkdocs.yml` and move it to `share/doc/revng/developer-manual/xxx.md`. It has been a while that I wanted to move stuff there.

This way we have no ToC issues, mkdocs will create it for us.

What do you think?

Reply:
Perfect! I didn't do it because, hey, this PR is already stretching the scope to its limits, and documentation structuring is something we'll definitely have to take a closer look at once we have more of these random writeups piled up.

I didn't want to add anything more to that already crowded `markdown` in `yaml` in `c++ header` monstrosity we had, but giving this its place is even better than that.

That being said, I don't think this should fall under the manual category 🤔.

`developer-reference/abi-definition.md`? `developer-manual/reference/abi-definition.md`?

Reply:
Mmh, maybe for now put it under `references`. Then, once we'll create `developer-manual/` we'll split the references under  `user-manual/` and `developer-manual/`.

Also, one day this stuff might become part of the model, so not too bad to put it a more user facing area.
Let's put this in `references/`.

Reply:
Done.